cleverhans/utils_tf.py

```diff
@@ -300,3 +302,42 @@ def model_argmax(sess, x, predictions, samples):
         return np.argmax(probabilities)
     else:
         return np.argmax(probabilities, axis=1)
+
+
+def l2_normalize(x, dim, epsilon=1e-12, name=None):
+    """Helper function that normalizes vector to unit norm
+    :param x: the input placeholder
+    :param axis: axis to normalizes
+    :param epsilon: stabilizes division
+    :return: the l2 normalized vector
+    """
+    with tf.name_scope(name, "l2_normalize", [x]) as name:
+        x = tf.convert_to_tensor(x, name="x")
+        x /= (epsilon + tf.reduce_max(tf.abs(x), dim, keep_dims=True))
+        square_sum = tf.reduce_sum(tf.square(x), dim, keep_dims=True)
+        x_inv_norm = tf.rsqrt(np.sqrt(epsilon) + square_sum)
+        return tf.multiply(x, x_inv_norm, name)
+
+
+def normalize_perturbation(x, name=None):
```

Comment:
Is it necessary to have two functions `l2_normalize` and `normalize_perturbation`? It looks like they could be merged.

If we keep the two, could we find a more descriptive name than `normalize_perturbation` to indicate that it uses the l2 norm?